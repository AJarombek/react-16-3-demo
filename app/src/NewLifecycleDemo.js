/**
 * Describe the new lifecycle methods introduced in React 16.3.
 * @author Andrew Jarombek
 * @since 1/17/2020
 */

import React from 'react';
import FeaturePage from './FeaturePage';
import { AJCodeSnippet } from 'jarombek-react-components';
import DerivedFromPropsWrapper from './newlifecycle/DerivedFromPropsWrapper';

const NewLifecycleDemo = () => {
  return (
    <FeaturePage>
      <h1>New Lifecycle Methods</h1>
      <div className="demo-body">
        <p>
          Last year I wrote an article about <a
          href="https://jarombek.com/blog/jan-19-2019-react-lifecycles">lifecycle
          methods in React</a>. It turns out my article was outdated upon arrival, since the
          React team began a multi-year process of deprecating existing lifecycles and adding
          new ones. In React 16.3, two new lifecycle methods were introduced -
          <code>getDerivedStateFromProps()</code> and <code>getSnapshotBeforeUpdate()</code>.
        </p>
        <p>
          It was also announced that three lifecycle methods will be slowly phased out over the
          coming releases. In React 17 (still not released at the time I'm writing this - Jan.
          17th, 2020) <code>componentWillMount</code>, <code>componentWillReceiveProps</code>,
          and <code>componentWillUpdate</code> will be removed. Legacy code can simply rename
          these methods to <code>UNSAFE_componentWillMount</code>,
          <code>UNSAFE_componentWillReceiveProps</code>, and
          <code>UNSAFE_componentWillUpdate</code> instead of being refactored. These
          <code>UNSAFE</code> prefixed methods were introduced in React 16.3 in case developers
          want to rename existing methods before React 17 is released.
        </p>
        <p>
          In my opinion there were some confusing aspects to some of these deprecated lifecycle
          methods. For example, <code>componentWillMount()</code> is invoked before
          <code>componentDidMount()</code>. However, API calls should occur in
          <code>componentDidMount()</code>, going against conventional logic that API calls
          should be initialized as soon as possible<sup>8</sup>. Hopefully removing these
          lifecycle methods will help simplify things.
        </p>
        <p>
          Once React 17 is released, the class component lifecycles will be:
        </p>
        <p>
          <code className="code-listing">constructor()</code> (Included although technically not
          a lifecycle method)
        </p>
        <p>
          <code className="code-listing">getDerivedStateFromProps(props, state)</code>
        </p>
        <p>
          <code className="code-listing">render()</code>
        </p>
        <p>
          <code className="code-listing">componentDidMount()</code>
        </p>
        <p>
          <code className="code-listing">shouldComponentUpdate()</code>
        </p>
        <p>
          <code className="code-listing">getSnapshotBeforeUpdate(prevProps, prevState)</code>
        </p>
        <p>
          <code className="code-listing">
            componentDidUpdate(prevProps, prevState, snapshot)
          </code>
        </p>
        <p>
          <code className="code-listing">componentWillUnmount()</code>
        </p>
        <p>
          <code className="code-listing">getDerivedStateFromError(error)</code>
        </p>
        <p>
          <code className="code-listing">componentDidCatch(error, info)</code>
        </p>
        <p>
          Note that <code>componentDidCatch()</code> was released in React 16.0 and
          <code>getDerivedStateFromError()</code> in React 16.6. Let's take a look at
          <code>getDerivedStateFromProps()</code> and <code>getSnapshotBeforeUpdate()</code>
          specifically since they were released with React 16.3.
        </p>
        <p>
          <code>getDerivedStateFromProps()</code> replaces the soon to be deprecated
          <code>componentWillReceiveProps()</code> lifecycle method. The main difference
          between the two is that <code>componentWillReceiveProps()</code> allows for side
          effects while <code>getDerivedStateFromProps()</code> does not<sup>9</sup>. What this
          means is that <code>getDerivedStateFromProps()</code> does not have access to the
          component instance while <code>componentWillReceiveProps()</code> does<sup>10</sup>.
          This is due to <code>getDerivedStateFromProps()</code> being a static method. The
          React team made this change because providing developers access to the component
          instance in <code>componentWillReceiveProps()</code> led to code that didn't match the
          lifecycle methods intended use.
        </p>
        <p>
          <code>getDerivedStateFromProps()</code> is invoked before a component renders. Based
          on the methods name, its main purpose is to update the state, which is derived from
          the props.
        </p>
        <DerivedFromPropsWrapper/>
      </div>
    </FeaturePage>
  );
};

export default NewLifecycleDemo;
