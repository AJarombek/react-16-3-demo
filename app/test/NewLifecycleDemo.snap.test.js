/**
 * Snapshot test for the {@link ../src/NewLifecycleDemo} component.
 * @author Andrew Jarombek
 * @since 1/30/2020
 */

import React from 'react';
import NewLifecycleDemo from '../src/NewLifecycleDemo';
import renderer from 'react-test-renderer';

jest.mock('react-router-dom', () => {
  const historyObj = {
    push: jest.fn()
  };

  return {
    ...jest.requireActual('react-router-dom'),
    useHistory: () => historyObj
  }
});

it('renders correctly', () => {
  const tree = renderer.create(<NewLifecycleDemo />).toJSON();
  expect(tree).toMatchSnapshot();
});
