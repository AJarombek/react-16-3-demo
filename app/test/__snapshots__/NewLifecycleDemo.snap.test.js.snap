// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="ajNavTextCircle-0-2-1 textRight-0-2-2"
  >
    <div
      className="aj-nav-circle ajNavCircle-0-2-5 ajNavCircle-d0-0-2-13 active-0-2-6"
      onClick={[Function]}
    >
      <figure>
        <img
          className="nav-left left-0-2-11 img-0-2-8 img-d1-0-2-14"
          src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAAAVlpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IlhNUCBDb3JlIDUuNC4wIj4KICAgPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4KICAgICAgPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIKICAgICAgICAgICAgeG1sbnM6dGlmZj0iaHR0cDovL25zLmFkb2JlLmNvbS90aWZmLzEuMC8iPgogICAgICAgICA8dGlmZjpPcmllbnRhdGlvbj4xPC90aWZmOk9yaWVudGF0aW9uPgogICAgICA8L3JkZjpEZXNjcmlwdGlvbj4KICAgPC9yZGY6UkRGPgo8L3g6eG1wbWV0YT4KTMInWQAAAjVJREFUeAHt21tqhDAYhmFvu4ki3VnvuugpMgvp4Q/lAxuiJh5iDq8Q0hkdTZ+3zkXBYWBDAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAIHhBYPDAqcZvtpSHjY+Di+p3xO8268+2Xg7SuBifNr4sfFtgyiGkLi5GF82nOHTxu4o8xjuZEQxhMRtHkOGu6KEYuiE3ClxVUIxZJgcZS0Id8p2kLUYzi85iLskUbbhQ0dcEkMXIook4uZLY2gJRJHE+pwlhpZAFEmE56wxtASiSOL/fEsMLYEokvibb42hpRCloBhEKTBG71GK+JpSBH/u7eur6BiK00uUKmL0EqWqGK1HqTJGq1GqjtFalCZitBKlqRi1R2kyRq1Rmo5RW5QuYtQSpasYpUfpMkapUbqOUVoUYqiIzXf/Q5IYsxj68a4oxFCBwJw7CjECEfy3ckUhhi+/8vrqKMRYwV/adVUUYiyJR7x/dhRiRKBvHXJWFGJsSSfsPxqFGAnYsYfujUKMWOEdx6VGIcYO5NSPxEYhRqrsgeNH++xkQw9N+rN77FiPHvv73OvJxmiD7USBrTslFMK997Sx+9nwE9ff5KlSoxAjw59BbBRiZIihS2xFIYakMs5LUYiRMYJ/KT8KMXyhG14rCjFuwF+65Gg73GBDAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAAEEEEAAAQQQQAABBBBAAIGSBX4BK4DEtcIi4HkAAAAASUVORK5CYII="
        />
      </figure>
    </div>
    <div
      className="description-0-2-4"
    >
      Home
    </div>
  </div>
  <h1>
    New Lifecycle Methods
  </h1>
  <div
    className="demo-body"
  >
    <p>
      Last year I wrote an article about 
      <a
        href="https://jarombek.com/blog/jan-19-2019-react-lifecycles"
      >
        lifecycle methods in React
      </a>
      . It turns out my article was outdated upon arrival, since the React team began a multi-year process of deprecating existing lifecycles and adding new ones. In React 16.3, two new lifecycle methods were introduced -
      <code>
        getDerivedStateFromProps()
      </code>
       and 
      <code>
        getSnapshotBeforeUpdate()
      </code>
      .
    </p>
    <p>
      It was also announced that three lifecycle methods will be slowly phased out over the coming releases. In React 17 (still not released at the time I'm writing this - Jan. 17th, 2020) 
      <code>
        componentWillMount
      </code>
      , 
      <code>
        componentWillReceiveProps
      </code>
      , and 
      <code>
        componentWillUpdate
      </code>
       will be removed. Legacy code can simply rename these methods to 
      <code>
        UNSAFE_componentWillMount
      </code>
      ,
      <code>
        UNSAFE_componentWillReceiveProps
      </code>
      , and
      <code>
        UNSAFE_componentWillUpdate
      </code>
       instead of being refactored. These
      <code>
        UNSAFE
      </code>
       prefixed methods were introduced in React 16.3 in case developers want to rename existing methods before React 17 is released.
    </p>
    <p>
      In my opinion there were some confusing aspects to some of these deprecated lifecycle methods. For example, 
      <code>
        componentWillMount()
      </code>
       is invoked before
      <code>
        componentDidMount()
      </code>
      . However, API calls should occur in
      <code>
        componentDidMount()
      </code>
      , going against conventional logic that API calls should be initialized as soon as possible
      <sup>
        8
      </sup>
      . Hopefully removing these lifecycle methods will help simplify things.
    </p>
    <p>
      Once React 17 is released, the class component lifecycles will be:
    </p>
    <p>
      <code
        className="code-listing"
      >
        constructor()
      </code>
       (Included although technically not a lifecycle method)
    </p>
    <p>
      <code
        className="code-listing"
      >
        getDerivedStateFromProps(props, state)
      </code>
    </p>
    <p>
      <code
        className="code-listing"
      >
        render()
      </code>
    </p>
    <p>
      <code
        className="code-listing"
      >
        componentDidMount()
      </code>
    </p>
    <p>
      <code
        className="code-listing"
      >
        shouldComponentUpdate()
      </code>
    </p>
    <p>
      <code
        className="code-listing"
      >
        getSnapshotBeforeUpdate(prevProps, prevState)
      </code>
    </p>
    <p>
      <code
        className="code-listing"
      >
        componentDidUpdate(prevProps, prevState, snapshot)
      </code>
    </p>
    <p>
      <code
        className="code-listing"
      >
        componentWillUnmount()
      </code>
    </p>
    <p>
      <code
        className="code-listing"
      >
        getDerivedStateFromError(error)
      </code>
    </p>
    <p>
      <code
        className="code-listing"
      >
        componentDidCatch(error, info)
      </code>
    </p>
    <p>
      Note that 
      <code>
        componentDidCatch()
      </code>
       was released in React 16.0 and
      <code>
        getDerivedStateFromError()
      </code>
       in React 16.6. Let's take a look at
      <code>
        getDerivedStateFromProps()
      </code>
       and 
      <code>
        getSnapshotBeforeUpdate()
      </code>
      specifically since they were released with React 16.3.
    </p>
    <h2>
      <code>
        getDerivedStateFromProps()
      </code>
    </h2>
    <p>
      <code>
        getDerivedStateFromProps()
      </code>
       replaces the soon to be deprecated
      <code>
        componentWillReceiveProps()
      </code>
       lifecycle method. The main difference between the two is that 
      <code>
        componentWillReceiveProps()
      </code>
       allows for side effects while 
      <code>
        getDerivedStateFromProps()
      </code>
       does not
      <sup>
        9
      </sup>
      . What this means is that 
      <code>
        getDerivedStateFromProps()
      </code>
       does not have access to the component instance while 
      <code>
        componentWillReceiveProps()
      </code>
       does
      <sup>
        10
      </sup>
      . This is due to 
      <code>
        getDerivedStateFromProps()
      </code>
       being a static method. The React team made this change because providing developers access to the component instance in 
      <code>
        componentWillReceiveProps()
      </code>
       led to code that didn't match the lifecycle methods intended use.
    </p>
    <p>
      <code>
        getDerivedStateFromProps()
      </code>
       is invoked before a component renders. Based on the methods name, its main purpose is to update the state, which is derived from the props.
    </p>
    <p>
      <code>
        getDerivedStateFromProps()
      </code>
       is called during the initial component render and each subsequent re-render.  It takes two arguments, the state and props of the component. Its expected to return either an object with properties to update the component state or
      <code>
        null
      </code>
       if the component state should remain unchanged.
    </p>
    <p>
      The following example consists of two components -
      <code>
        DerivedFromPropsWrapper
      </code>
       and 
      <code>
        DerivedFromProps
      </code>
      .  The purpose of
      <code>
        DerivedFromPropsWrapper
      </code>
       is to pass a prop down to
      <code>
        DerivedFromProps
      </code>
      .  The 
      <code>
        DerivedFromProps
      </code>
       component has a
      <code>
        secretCode
      </code>
       state property which is derived from its boolean
      <code>
        show
      </code>
       prop.
    </p>
    <div
      className="aj-code-snippet"
    >
      <code
        className="javascript hljs-0-2-15"
        title="JavaScript"
      >
        import React, { useState } from 'react';
import DerivedFromProps from './DerivedFromProps';
import { AJContainedButton } from 'jarombek-react-components';

const DerivedFromPropsWrapper = () =&gt; {
  const [show, setShow] = useState(false);

  return (
    &lt;div className="derived-from-props-wrapper"&gt;
      &lt;p&gt;&lt;code&gt;getDerivedStateFromProps()&lt;/code&gt; Demonstration&lt;/p&gt;
      &lt;AJContainedButton onClick={() =&gt; setShow(!show)}&gt;
        {show ? 'Show Secret Message' : 'Hide Secret Message'}
      &lt;/AJContainedButton&gt;
      &lt;DerivedFromProps show={show}/&gt;
    &lt;/div&gt;
  );
};

      </code>
    </div>
    <div
      className="aj-code-snippet"
    >
      <code
        className="javascript hljs-0-2-15"
        title="JavaScript"
      >
        import React from 'react';
import PropTypes from 'prop-types';

class DerivedFromProps extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      secretCode: 0
    };
  }

  static getDerivedStateFromProps(props, state) {
    // Assert that this lifecycle method does not have access to the component instance.
    console.assert(this, undefined);

    if (props.show) {
      return { secretCode: 1231 };
    } else {
      return { secretCode: 0 };
    }
  }

  render() {
    return (
      &lt;div className="derived-from-props"&gt;
        { this.state.secretCode === 1231 ?
          &lt;p className="secret-code"&gt;
            You have a beautiful heart.
          &lt;/p&gt;
          :
          &lt;div className="secret-code-classified"&gt;
            &lt;p&gt;CLASSIFIED&lt;/p&gt;
          &lt;/div&gt;
        }
      &lt;/div&gt;
    )
  }
}

DerivedFromProps.propTypes = {
  show: PropTypes.bool.isRequired
};

      </code>
    </div>
    <p>
      Here are the rendered components:
    </p>
    <div
      className="derived-from-props-wrapper"
    >
      <p>
        <code>
          getDerivedStateFromProps()
        </code>
         Demonstration
      </p>
      <div
        className="ajBaseButton-0-2-19 ajBaseButtonEnabled-0-2-20 aj-contained-button ajContainedButton-0-2-16 ajContainedButtonEnabled-0-2-17"
        onClick={[Function]}
      >
        <button
          disabled={false}
          type="button"
        >
          Hide Secret Message
        </button>
      </div>
      <div
        className="derived-from-props"
      >
        <div
          className="secret-code-classified"
        >
          <p>
            CLASSIFIED
          </p>
        </div>
      </div>
    </div>
    <p>
      Although this example helps demonstrate the basic functionality of
      <code>
        getDerivedStateFromProps()
      </code>
      , in production code 
      <code>
        DerivedFromProps
      </code>
      should be refactored to be a stateless component.  Here is the code for
      <code>
        DerivedFromProps
      </code>
       after proper refactoring:
    </p>
    <div
      className="aj-code-snippet"
    >
      <code
        className="javascript hljs-0-2-15"
        title="JavaScript"
      >
        import React from 'react';
import PropTypes from 'prop-types';

const DerivedFromPropsRefactored = ({show}) =&gt; {
  return (
    &lt;div className="derived-from-props"&gt;
      { show ?
        &lt;p className="secret-code"&gt;
          You have a beautiful heart.
        &lt;/p&gt;
        :
        &lt;div className="secret-code-classified"&gt;
          &lt;p&gt;CLASSIFIED&lt;/p&gt;
        &lt;/div&gt;
      }
    &lt;/div&gt;
  )
};

DerivedFromPropsRefactored.propTypes = {
  show: PropTypes.bool.isRequired
};

      </code>
    </div>
    <h2>
      <code>
        getSnapshotBeforeUpdate()
      </code>
    </h2>
    <p>
      <code>
        getSnapshotBeforeUpdate()
      </code>
       is a new lifecycle method used in conjunction with
      <code>
        componentDidUpdate()
      </code>
      .  A combination of the two can be used as a replacement for the deprecated 
      <code>
        componentWillUpdate()
      </code>
       method.
    </p>
    <p>
      <code>
        getSnapshotBeforeUpdate()
      </code>
       and 
      <code>
        componentDidUpdate()
      </code>
       are both invoked right after 
      <code>
        render()
      </code>
       executes.  However,
      <code>
        getSnapshotBeforeUpdate()
      </code>
       is invoked before changes to the virtual DOM are committed to the DOM
      <sup>
        11
      </sup>
      .  On the other hand, 
      <code>
        componentDidUpdate()
      </code>
      is invoked after changes to the virtual DOM are reflected in the DOM.
    </p>
    <p>
      The return value of 
      <code>
        getSnapshotBeforeUpdate()
      </code>
       is passed to the
      <code>
        snapshot
      </code>
       parameter in 
      <code>
        componentDidUpdate()
      </code>
      .  The typical use case of 
      <code>
        getSnapshotBeforeUpdate()
      </code>
       is to extract information from the DOM API before DOM changes take place and pass that information along to the moment after DOM changes occur.  I'm honestly still trying to think of scenarios where this would be useful.  The React blog uses a chat application as an example, and most other online sources simply copy and rephrase it
      <sup>
        12
      </sup>
      .
    </p>
    <p>
      The example of 
      <code>
        getSnapshotBeforeUpdate()
      </code>
       shown below isn't useful as a real world scenario, however it does show the different moving pieces in the lifecycle method.
    </p>
    <div
      className="snapshot-before-update"
    >
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "aqua",
          }
        }
      >
        AQUA
      </button>
      <p>
        Previous Color: 
        <strong>
          none
        </strong>
      </p>
    </div>
    <div
      className="aj-code-snippet"
    >
      <code
        className="javascript hljs-0-2-15"
        title="JavaScript"
      >
        import React, {createRef} from 'react';

class SnapshotBeforeUpdate extends React.Component {

  constructor(props) {
    super(props);

    this.colors = [
      'aqua', 'aquamarine', 'lightskyblue'
    ];

    this.buttonRef = createRef();
    this.state = {
      clicks: 0,
      prevColor: 'none'
    };
  }

  onButtonClick() {
    this.setState({clicks: this.state.clicks + 1})
  }

  getSnapshotBeforeUpdate(prevProps, prevState) {
    return {
      color: this.buttonRef.current.style.backgroundColor
    }
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    if (snapshot !== null && snapshot.color !== this.buttonRef.current.style.backgroundColor) {
      this.setState({
        prevColor: snapshot.color
      });
    }
  }

  render() {
    return (
      &lt;div className="snapshot-before-update"&gt;
        &lt;button
          ref={this.buttonRef}
          onClick={() =&gt; this.onButtonClick()}
          style={{backgroundColor: this.colors[this.state.clicks % 3]}}&gt;

          {this.colors[this.state.clicks % 3].toUpperCase()}
        &lt;/button&gt;
        &lt;p&gt;Previous Color: &lt;strong&gt;{this.state.prevColor}&lt;/strong&gt;&lt;/p&gt;
      &lt;/div&gt;
    )
  }
}

      </code>
    </div>
  </div>
</div>
`;
